### DISCUSSION TITLE
Reduce space between headers and body text

### DISCUSSION THREAD

**Discussion by deminimis** [squidfunk/mkdocs-material#7366] [Q&A] [answered]
A couple small questions. 

1) Is there a way to reduce space between the headers (both markdown and html headers)? Such as:
`.md-typeset h2 {
    margin-bottom: 5px;
}`

2) Is there a way to globally hide the ToC? The exact function as 
```
Hide:
  - toc
```
Without having to do it on every page, or write all headers in html. 

I am having trouble understanding the partials and blocks. When I added: custom_dir: overrides, even if there was no main.html or toc.html, it made both navigation on left side and toc on right disappear. I thought I got Giscus commenting to do something, but maybe I have to check after pushing to Github, because on mkdocs serve it gives me a small window of my website at the bottom. 

I tried adding some random things in extra.css that I used for my custom color theme but they didn't do anything. 
```
.hidden-toc {
  display: none;
}


.md-nav--primary {
	display: none
}

.md-nav--secondary {
	display: none
}
```

**Comment by kamilkrzyskow** [comment]
Hello @jonsnow231,
please be aware that it will be better overall if you learn how to work with CSS outside of mkdocs-material. You know, baby steps and building up knowledge before jumping straight into the deep see of customization, having said that let's see if I can help you:

> Is there a way to reduce space between the headers (both markdown and html headers)? Such as:

I'm not sure why you separate between Markdown and HTML headings 🤔 MkDocs processes the Markdown and this gets turned into HTML, and the web browser uses CSS to style HTML, not Markdown. You can use the browser web developer tools to see what is taking up space for a given element. In the example below the H3 tag we can see it has both margin-top and margin-bottom. You can modify the extra CSS for a single tag, or use the `:is` selector to modify multiple:
- https://developer.mozilla.org/en-US/docs/Web/CSS/margin
- https://developer.mozilla.org/en-US/docs/Web/CSS/:is
- https://caniuse.com/css-matches-pseudo

Vertical margin auto collapses, so if 2 adjacent elements have margin on the same side it will only behave as it was 1 margin (the bigger of the 2 if applicable).

<details>

![image](https://github.com/user-attachments/assets/8dea8ef1-80d8-4fe5-9e95-86f519fcc87a)

</details>

---

> Is there a way to globally hide the ToC? The exact function as

No, there is not, as seen in the templates the value for the hidden sidebar is taken directly from the `page.meta.hide` dict:
https://github.com/squidfunk/mkdocs-material/blob/4f8081c268d31bf74d546e600cadd0cff7dc89e8/src/templates/base.html#L316-L326

Overriding `base.html` is highly discouraged, so you either use the [meta plugin](https://squidfunk.github.io/mkdocs-material/plugins/meta/) to edit the `page.meta` value for all pages in a directory, or create a [hook](https://www.mkdocs.org/user-guide/configuration/#hooks) to do the same thing as the plugin, but coded by hand, or use CSS to apply a `display: hidden/none`, but please be aware that in case of slow internet connection the sidebar navigation could be shown before the CSS files load.

---

> I am having trouble understanding the partials and blocks.

The entry point for every generated page is the `main.html` ([unless you change it](https://squidfunk.github.io/mkdocs-material/reference/#setting-the-page-template)), which by default extends the `base.html` prepared by the Material theme.
This `base.html` file defines `block`s that can be extended by the extending `main.html`. For example `htmltitle`:
https://github.com/squidfunk/mkdocs-material/blob/4f8081c268d31bf74d546e600cadd0cff7dc89e8/src/templates/base.html#L89-L98

If you override it in your `main.html`, you can replace the whole block with custom content, or using the `{{ super() }}` function you can load the content of the parent file that is being extended (`base.html`) and you can add your custom content before or after the original contents.

As for partials the program holds an internal mapping of available template files for example `partials/javascripts/base.html`:
https://github.com/squidfunk/mkdocs-material/blob/4f8081c268d31bf74d546e600cadd0cff7dc89e8/src/templates/base.html#L153-L154

By default it will look for that [`partials/javascripts/base.html`](https://github.com/squidfunk/mkdocs-material/blob/master/src/templates/partials/javascripts/base.html) file in the theme's templates, but with the `custom_dir: overrides` you can create a 1:1 duplicate file structure (for any path in the [templates](https://github.com/squidfunk/mkdocs-material/tree/master/material/templates)) and then it will prioritize using that file in your custom directory.

Simply put, follow the thread of execution from top to bottom, starting from `main.html`.

---

As for the  `.hidden-toc` rule, there is no such class in the default style sheet, so how are you setting in for the Markdown/HTML?
As for the `.md-nav--primary` rule, there should be no issues with this approach, [Specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity) is upheld, so make sure that your CSS file is actually being loaded. Once again the web developer tools of your browser are your friend in this case. You can see what is being loaded, and what is the contents of the files. There are tutorials online how to manage it.

**Comment by deminimis** [comment]
If you add the html rather than markdown headers, they don't show up in the ToC. I saw it in these discussions somewhere. Actually, while I was trying to find it again, I stumbled on one way to hide the ToC globally: https://github.com/squidfunk/mkdocs-material/issues/478#issuecomment-578723876

```markdown_extensions:
    - toc:
        toc_depth : "1-1"
```

But I have no idea how to add those blocks/snippets still. I tried following the [customization](https://squidfunk.github.io/mkdocs-material/customization/#additional-stylesheets) to add something like:

```{% extends "base.html" %}

{% block content %}
    {{ super() }}
.md-typeset h2 {
    font-size: 1.5625em;
    line-height: 1.4;
    margin: 4em 0 .64em;
    margin-top: 1.6em;
    margin-right: 0px;
    margin-bottom: 0.64em;
    margin-left: 0px;
}
{% endblock %}
```

Even trying to put it in main.html, content.html, extra.css, and splitting up the .md in extra.css and the other in the other two, but I couldn't figure it out. It made no changes at all, just like the .md-nav--secondary. So I couldn't try to figure out specificity with the inspect tool, because there were no changes to it at all. 

The extra.css is working for my custom theme where I just changed colors, but nothing else.

**Reply by kamilkrzyskow** [reply]
Oh, that's what you meant with hiding the ToC, this is a bit different, but overall nice ideas to hack it, I didn't event consider them, so thanks.
Both of your suggestions prevent the toc items from rendering on the page, but the sidebar should stil be present and take up width.

I'm not sure what that customization tries to do, as adding CSS style sheet needs to be in an extra file not in the content block 🤔 
Extension blocks like `block content` can only be in the extending file, so `main.html` or another entry point, not in partials.

Please provide a minimal example of the project, so that we're on the same page of what files are being used.
- https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/

As you're working with customizations remember to disable:
- https://squidfunk.github.io/mkdocs-material/plugins/info/#config.archive_stop_on_violation

**Comment by deminimis** [comment]
So then if I wanted to add this css:
```
.md-typeset h2 {
    font-size: 1.5625em;
    line-height: 1.4;
    margin: 4em 0 .64em;
    margin-top: 1.6em;
    margin-right: 0px;
    margin-bottom: 0.64em;
    margin-left: 0px;
}
```

What would be the correct place to add it in a new mkdocs folder? It sounds like I am just not putting it in correctly.

**Reply by kamilkrzyskow** [reply, marked as answer]
https://squidfunk.github.io/mkdocs-material/customization/#additional-css
Inside the extra.css file just like the colors.

**Reply by deminimis** [reply]
Ahh, okay, thank you. I got it to work by placing inside the 
```
[data-md-color-scheme="customcolor"] {
}
```
I had placed outside thinking it would work globally. So now I want to change the same heading margins on the "slate" and "default" themes. Would it be better to just create custom css copying those to do that? I tried plugging slate and default into that data-md-color-scheme=, but it didn't do anything.

**Reply by kamilkrzyskow** [reply]
It should work globally without any attribute selectors, also a catch-all attribute selector looks like this `[data-md-color-scheme]` without the value part, should match both slate/default. 
I would say stick to 1 extra.css file to optimize web requests and caching, if the changes are global for the whole site.

EDIT: Perhaps you make some mistakes in the syntax, which aborts the loading process of the CSS file. Try running the CSS in a online validator.
